src/charm.py

```diff
@@ -50,42 +82,22 @@ def _on_config_changed(self, _) -> None:
             self.model.unit.status = ops.BlockedStatus(exc.msg)
             return
         if self.model.unit.is_leader():
-            for relation in self.model.relations["saml"]:
-                relation.data[self.model.app].update(self.dump_saml_data())
+            for relation in self.relations:
+                self._update_relation_data(relation, self.get_saml_data())
         self.unit.status = ops.ActiveStatus()
 
-    def dump_saml_data(self) -> Dict[str, str]:
-        """Dump the charm state in the format expected by the relation.
+    def get_saml_data(self) -> saml.SamlRelationData:
+        """Get relation data.
 
         Returns:
-            Dict containing the IdP details.
+            SamlRelationData containing the IdP details.
         """
-        result = {
-            "entity_id": self._saml_integrator.entity_id,
-            "metadata_url": self._saml_integrator.metadata_url,
-            "x509certs": ",".join(self._saml_integrator.certificates),
-        }
-        for endpoint in self._saml_integrator.endpoints:
-            http_method = endpoint.binding.split(":")[-1].split("-")[-1].lower()
-            lowercase_name = re.sub(r"(?<!^)(?=[A-Z])", "_", endpoint.name).lower()
-            prefix = f"{lowercase_name}_{http_method}_"
-            result[f"{prefix}url"] = endpoint.url
-            result[f"{prefix}binding"] = endpoint.binding
-            if endpoint.response_url:
-                result[f"{prefix}response_url"] = endpoint.response_url
-        return result
-
-    def _on_saml_relation_created(self, event: ops.RelationCreatedEvent):
-        """Handle a change to the saml relation.
-
-        Populate the event data.
-
-        Args:
-            event: Event triggering the relation-created hook for the relation.
-        """
-        if not self.model.unit.is_leader():
-            return
-        event.relation.data[self.model.app].update(self.dump_saml_data())
+        return saml.SamlRelationData(
+            entity_id=self._charm_state.entity_id,
+            metadata_url=self._charm_state.metadata_url,
```

Comment:
If it were initialised in the constructor, it would always fail on testing due to the pydantic validations not passing. The testing framework doesn't reinstantiate the charm when a config changed is emitted, as oposed to the behaviour in ops

Reply:
Franco does have a point, the charm state should be initialised in `__init__`. Why do the tests fail if that is the case?

Reply:
I was able to workaround the harness limitations  and changed this as part of https://github.com/canonical/saml-integrator-operator/pull/14. I'm closing the conversation